src/async_wrap.cc

```diff
@@ -670,6 +674,10 @@ void AsyncWrap::AsyncReset(Local<Object> resource, double execution_async_id,
                                                      : execution_async_id;
   trigger_async_id_ = env()->get_default_trigger_async_id();
 
+  if (resource != object()) {
+    resource_.Reset(env()->isolate(), resource);
```

Comment:
This is still very much a footgun, imo. I think it currently happens to work out in the cases in which we set a custom resource, but that’s easy to get wrong when expanding this to other resource types. (And I think for HTTP `Parser`s it also only works because it’s masked by another potential memory leak).

I don’t want to keep this PR from being landed, as important as it is, but I’d at least leave a comment about it.

```suggestion
    // TODO(addaleax): Using a strong reference here makes it very easy to
    // introduce memory leaks. Move away from using a strong reference.
    resource_.Reset(env()->isolate(), resource);
```

Reply:
I've added the note so this should be in a ready state again.

I'll continue to think about if there's a way to do this without the strong reference. I've said before that my intent is to start moving as much of async_hooks to the JavaScript side as possible, and I think moving all the resource-tracking stuff would eliminate this.

Reply:
> I'll continue to think about if there's a way to do this without the strong reference.

Fwiw, I did point out one earlier: https://github.com/nodejs/node/pull/30959#discussion_r374780123

My intention with the `TODO` comment is basically to implement that once this PR has landed.

Reply:
Ah missed that. It's getting hard to keep track of all the feedback. :sweat_smile: 

I don't really understand exactly what you're suggesting there though. Perhaps we can discuss it separately, if you have some time at some point? I intend to continue working on more async_hooks improvements. :smile:

Reply:
The idea (without having tried it myself) would be to:

1. Turn this `resource_` back into a weak reference
2. `object()->Set(env->context(), env->owner_symbol(), resource)`

in order to create a “strong” reference from the AsyncWrap binding object that the GC can observe (as opposed to the current state in which it cannot observe the logical connection between the two).

If the `Set()` call is too slow, one alternative that could be investigated is using a second internal field on `AsyncWrap` subclasses that could hold the owner (= public resource) object.